README.md
# Voice Skill Management System (Core)

> A framework aiming at the rapid development of custom interactions (skills) for voice assistant systems.
 
## About
This library is part of the Voice Skill Management System, or VSMS for short. 

VSMS is a skill development framework for voice assistants such as Amazon Voice Service. The intention is to speed up 
development by incorporating boilerplate code into a framework. VSMS is preconfigured to run multiple skills
as well as serving HTML content.    

## Usage
Please head over to the [VSMS-Skeleton](https://github.com/internetofvoice/vsms-skeleton)
to start developing your voice assistant skills. 

VSMS-Skeleton is an application skeleton based on this library and the Slim Framework 
which will help you to get started with voice assistant development.

## Development
VSMS aims to interact with multiple voice assistant systems, so it's not only nailed down to Amazon Voice Service. 
With more services popping up, VSMS will be extended to support those too. The idea is to have a single environment 
to handle all your voice assistant development needs without writing duplicate code.

## Concluding words
You are welcome to contribute to this project, and so are your pull requests.

Found a bug or have a feature request? Please open a new issue and let us know.


CHANGELOG.md
# Changelog

Changes are documented in reverse chronological order.

## Version 1.4.0 (2019-05-30)
* Switch to libvoice 1.3.0 and PHPUnit 7     

## Version 1.3.3 (2019-04-01)
* Extend SkillHelper::extractAmazonDate to recognize XXXX-XX-01 and 2019-01-XX date values     

## Version 1.3.2 (2019-03-11)
* Extend LogHelper::logRequest to optionally not include request headers  

## Version 1.3.1 (2019-02-19)
* Add support for future dates in SkillHelper::convertDateTimeToHuman()  

## Version 1.3.0 (2019-02-17)
* Support variants in TranslationHelper (multiple translations, pick one by random)  

## Version 1.2.4 (2018-12-16)
* Switch to [analog-stable](https://github.com/jbroadway/analog) library

## Version 1.2.3 (2018-01-21)
* Fix bug in extraction of AMAZON.DATE slot type 
* Extend exception handling
* Improve test coverage

## Version 1.2.2 (2017-12-19)
* Extend exception handling

## Version 1.2.1 (2017-10-11)
* Add separate prefix for today / yesterday to SkillHelper->convertDateTimeToHuman()

## Version 1.2.0 (2017-10-11)
* Switch to [LibVoice](https://github.com/internetofvoice/libvoice) library

## Version 1.0.1 (2017-09-27)
* Add exception handling

## Version 1.0.0 (2017-09-11)
* First public release


test/Fixtures/settings.php
<?php

/**
 * Settings fixtures
 *
 * @author	Alexander Schmidt <a.schmidt@internet-of-voice.de>
 * @license http://opensource.org/licenses/MIT
 */

return [
	'settings' => [
		'displayErrorDetails'    => true,
		'addContentLengthHeader' => false,
		'validateTimestamp'      => false,
		'validateCertificate'    => false,
		'environment'            => 'dev-test',

		'auto_init' => ['logger', 'translator', 'skillHelper'],

		'locale_default' => 'en-US',
		'locales'        => ['de-DE', 'en-US', 'en-GB'],
	],
];


test/Fixtures/MockService.php
<?php

namespace Tests\InternetOfVoice\VSMS\Core\Fixtures;

use InternetOfVoice\VSMS\Core\Service\AbstractService;

/**
 * Class MockService
 *
 * @author  Alexander Schmidt <a.schmidt@internet-of-voice.de>
 * @license http://opensource.org/licenses/MIT
 */
class MockService extends AbstractService {
}


test/Helper/LogHelperTest.php
<?php

namespace Tests\InternetOfVoice\VSMS\Core\Helper;

use Analog\Handler\LevelName;
use Analog\Handler\Variable;
use InternetOfVoice\VSMS\Core\Helper\LogHelper;
use Tests\InternetOfVoice\VSMS\Core\Controller\ControllerTestCase;

/**
 * Class LogHelperTest
 *
 * @author  Alexander Schmidt <a.schmidt@internet-of-voice.de>
 * @license http://opensource.org/licenses/MIT
 */
class LogHelperTest extends ControllerTestCase {
    /**
     * testLogger
     */
    public function testLogger() {
        $log    = '';
        $logger = new LogHelper;
        $logger->format('%s - %s - %s - %s' . PHP_EOL);
        $logger->handler(LevelName::init(Variable::init($log)));

        $logger->info('A test log entry.');
        $this->assertContains('INFO', $log);
        $this->assertContains('A test log entry.', $log);
    }

    public function testRequestLogger() {
        $request = $this->createRequest('POST', '/test/url?param=value', [], json_encode(['key' => 'value', 'key2' => 'value2']));
        $log     = '';
        $logger  = new LogHelper;
        $logger->format('%s - %s - %s - %s' . PHP_EOL);
        $logger->handler(LevelName::init(Variable::init($log)));
        $logger->setMask(['key']);
        $logger->logRequest($request, ['extra-key' => 'extra-value']);

        $this->assertContains('DEBUG - POST LogHelperTest::testRequestLogger {"extra-key":"extra-value"}', $log);
	    $this->assertContains('DEBUG - Query params', $log);
        $this->assertContains('DEBUG - Header', $log);
        $this->assertContains('DEBUG - Body: {"key":"***","key2":"value2"}', $log);
    }

	public function testRequestLoggerNoHeader() {
		$request = $this->createRequest('POST', '/test/url?param=value', [], json_encode(['key' => 'value', 'key2' => 'value2']));
		$log     = '';
		$logger  = new LogHelper;
		$logger->format('%s - %s - %s - %s' . PHP_EOL);
		$logger->handler(LevelName::init(Variable::init($log)));
		$logger->setMask(['key']);
		$logger->logRequest($request, ['extra-key' => 'extra-value'], false);

		$this->assertContains('DEBUG - POST LogHelperTest::testRequestLoggerNoHeader {"extra-key":"extra-value"}', $log);
		$this->assertContains('DEBUG - Query params', $log);
		$this->assertNotContains('DEBUG - Header', $log);
		$this->assertContains('DEBUG - Body: {"key":"***","key2":"value2"}', $log);
	}
}


src/Helper/LogHelper.php
<?php

namespace InternetOfVoice\VSMS\Core\Helper;

use Analog\Logger;

/**
 * Class LogHelper
 *
 * @author  Alexander Schmidt <a.schmidt@internet-of-voice.de>
 * @license http://opensource.org/licenses/MIT
 */
class LogHelper extends Logger {
    /** @var array $mask */
    protected $mask = [];

    /**
     * Get mask
     *
     * @return  array
     * @access	public
     * @author	a.schmidt@internet-of-voice.de
     */
    public function getMask() {
        return $this->mask;
    }

    /**
     * Set mask
     *
     * @param   array   $mask   Keys, whose values should be masked out in logs (helpful to suppress sensitive data)
     * @access	public
     * @author	a.schmidt@internet-of-voice.de
     */
    public function setMask($mask) {
        $this->mask = $mask;
    }

    /**
     * Log server request
     *
     * @param 	\Slim\Http\Request	    $request 	    Request object
     * @param 	array 					$extra          Additional log data
     * @param 	bool 					$includeHeader  Whether to include header data
     * @access	public
     * @author	a.schmidt@internet-of-voice.de
     */
    public function logRequest($request, $extra = [], $includeHeader = true) {
        $extra = count($extra) ? ' ' . json_encode($extra) : '';

        // Caller info
        $trace = debug_backtrace(false, 2);
        if(isset($trace[1])) {
            $caller = substr(strrchr($trace[1]['class'], '\\'), 1) . '::' .  $trace[1]['function'];
            $this->debug($request->getMethod() . ' ' . $caller . $extra);
        }

        // Parameters, if applicable
        $params = $request->getQueryParams();
        if(!empty($params)) {
            $this->debug('Query params: ' . json_encode($params));
        }

        // Request headers, if applicable
	    if($includeHeader) {
		    $headers = array_map(function($element) {
			    return implode(', ', $element);
		    }, $request->getHeaders());

		    $this->debug('Header: ' . json_encode($headers));
	    }

        // Request body, if applicable
        $body = $request->getParsedBody();
        if(!is_null($body)) {
            if(is_array($body)) {
                // Apply logger mask to all nested elements
                $mask = array();
                foreach($this->getMask() as $key) {
                    array_push($mask, $key);
                }

                array_walk_recursive($body, function(&$value, $key) use ($mask) {
                    if(in_array($key, $mask)) {
                        $value = '***';
                    }
                });
            }

            $this->debug('Body: ' . json_encode($body));
        }
    }
}


src/Service/AbstractService.php
<?php

namespace InternetOfVoice\VSMS\Core\Service;

use InternetOfVoice\VSMS\Core\Helper\LogHelper;
use InternetOfVoice\VSMS\Core\Helper\SkillHelper;
use InternetOfVoice\VSMS\Core\Helper\TranslationHelper;

/**
 * Class AbstractService
 *
 * @author  Alexander Schmidt <a.schmidt@internet-of-voice.de>
 * @license http://opensource.org/licenses/MIT
 */
abstract class AbstractService
{
    /** @var LogHelper $logger */
    protected $logger;

    /** @var skillHelper $skillHelper */
    protected $skillHelper;

    /** @var TranslationHelper $translator */
    protected $translator;


    /**
     * @param   LogHelper           $logger
     * @access  public
     * @author  a.schmidt@internet-of-voice.de
     */
    public function setLogger($logger) {
        $this->logger = $logger;
    }

    /**
     * @param   SkillHelper         $skillHelper
     * @access  public
     * @author  a.schmidt@internet-of-voice.de
     */
    public function setSkillHelper($skillHelper) {
        $this->skillHelper = $skillHelper;
    }

    /**
     * @param   TranslationHelper   $translator
     * @access  public
     * @author  a.schmidt@internet-of-voice.de
     */
    public function setTranslator($translator) {
        $this->translator = $translator;
    }
}


src/Helper/TranslationHelper.php
<?php

namespace InternetOfVoice\VSMS\Core\Helper;

/**
 * Class TranslationHelper
 *
 * @author  Alexander Schmidt <a.schmidt@internet-of-voice.de>
 * @license http://opensource.org/licenses/MIT
 */
class TranslationHelper {
    /** @var array $locales  */
    protected $locales;

    /** @var string $locale_default  */
    protected $locale_default;

    /** @var array $messages  */
    protected $messages;

    /** @var string locale */
    protected $locale;

    /** @var string language */
    protected $language;


 	/**
	 * Constructor
	 *
     * @param   array       $locales            Supported locales
     * @param   string      $locale_default     Default locale
     * @access	public
	 * @author	a.schmidt@internet-of-voice.de
	 */
    public function __construct($locales, $locale_default) {
        $this->locales = $locales;
        $this->locale_default = $locale_default;
    }

    /**
     * Choose Locale
     *
     * Try to match preferred and supported locales/languages
     *
     * @param   string  $preferred  Preferred locales or languages, e.g. Accept-Language header
     * @return  bool
     * @access  public
     * @author  a.schmidt@internet-of-voice.de
     */
    public function chooseLocale($preferred) {
        $available_locales   = array_flip($this->locales);
        $available_languages = array();
        foreach($available_locales as $key => $value) {
            $available_languages[substr($key, 0, 2)] = $key;
        }

        preg_match_all('~([\w-]+)(?:[^,\d]+([\d.]+))?~', $preferred, $matches, PREG_SET_ORDER);

        $matched_locales = array();
        foreach($matches as $match) {
            $temp  = explode('-', $match[1]);
            $lang  = array_shift($temp);
            $value = isset($match[2]) ? (float)$match[2] : 1.0;

            // Full match (language_territory)?
            if(isset($available_locales[$match[1]]) && !isset($matched_locales[$match[1]])) {
                $matched_locales[$match[1]] = $value;
            }

            // Language match (without territory)?
            if(isset($available_languages[$lang]) && !isset($matched_locales[$available_languages[$lang]])) {
                $matched_locales[$available_languages[$lang]] = $value - 0.05;
            }
        }

        if(count($matched_locales)) {
            arsort($matched_locales);
            $this->setLocale(key($matched_locales));
            return true;
        } else {
            $this->setLocale($this->locale_default);
            return false;
        }
    }


    /**
     * Add translation
     *
     * @param   string      $path       Translation base path (up to, but without locale dir)
     * @param   string      $file       Translation file name (without .php extension)
     * @return  bool
     * @access	public
     * @author	a.schmidt@internet-of-voice.de
     */
    public function addTranslation($path, $file) {
        $filename = $path . '/' . $this->locale . '/' . $file . '.php';
        if(!is_readable($filename)) {
            return false;
        } else {
            /** @noinspection PhpIncludeInspection */
            $add_messages = require($filename);
            $this->messages = array_merge($this->messages, $add_messages);
            return true;
        }
    }


    /**
     * Translate text
     *
     * @return  mixed
     * @see     http://php.net/manual/en/function.sprintf.php
	 * @access	public
	 * @author	a.schmidt@internet-of-voice.de
     */
    public function t() {
        $args = func_get_args();
        if(count($args) < 1) {
            return false;
        }

        $message = array_shift($args);

        if(!isset($this->messages[$message])) {
	        // No translation available, return original string
            return vsprintf($message, $args);
        }

		$translation = $this->messages[$message];
        if(is_array($translation)) {
        	// Variations found, pick a translation by random
        	$translation = $translation[array_rand($translation)];
        }

        // Return translation with optional replacements
        return vsprintf($translation, $args);
    }

    /**
     * Get translation array (no content substitution)
     *
     * @param   string      $message
     * @return  mixed
     * @access	public
     * @author	a.schmidt@anschluss80.de
     */
    public function a($message) {
        return isset($this->messages[$message]) ? $this->messages[$message] : $message;
    }

    /**
     * Get locale
     *
     * @return  string
     * @access	public
     * @author	a.schmidt@internet-of-voice.de
     */
    public function getLocale() {
        return $this->locale;
    }

    /**
     * Set locale
     *
     * @param   string      $locale
     * @access	public
     * @author	a.schmidt@internet-of-voice.de
     */
    public function setLocale($locale) {
        $this->locale   = $locale;
        $this->language = substr($locale, 0, (strpos($locale, '-')));
        $this->messages = array();
    }

    /**
     * Get language
     *
     * @return  string
     * @access	public
     * @author	a.schmidt@internet-of-voice.de
     */
    public function getLanguage() {
        return $this->language;
    }

    /**
     * Set language
     *
     * @param   string      $language
     * @access	public
     * @author	a.schmidt@internet-of-voice.de
     */
    public function setLanguage($language) {
        $this->$language = $language;
    }
}


src/Controller/AbstractController.php
<?php

namespace InternetOfVoice\VSMS\Core\Controller;

use \Interop\Container\Exception\ContainerException;
use \Slim\Container;

/**
 * Class AbstractController
 *
 * @author  Alexander Schmidt <a.schmidt@internet-of-voice.de>
 * @license http://opensource.org/licenses/MIT
 */
abstract class AbstractController {
    /** @var \Slim\Container $container */
    protected $container;

    /** @var array $settings */
    protected $settings;

    /** @var \InternetOfVoice\VSMS\Core\Helper\LogHelper $logger */
    protected $logger;

    /** @var \InternetOfVoice\VSMS\Core\Helper\TranslationHelper $translator */
    protected $translator;

    /**
     * Constructor
     *
     * @param   \Slim\Container $container  Slim app container
     * @access  public
     * @throws  ContainerException
     * @author  a.schmidt@internet-of-voice.de
     */
    public function __construct(Container $container) {
        $this->container = $container;
        $this->settings  = $this->container->get('settings');

        if(in_array('logger', $this->settings['auto_init'])) {
            $this->logger = $this->container->get('logger');
        }

        if(in_array('translator', $this->settings['auto_init'])) {
            $this->translator = $this->container->get('translator');
            $this->translator->chooseLocale($container->request->getHeaderLine('Accept-Language'));
        }
    }
}


src/Helper/SkillHelper.php
<?php

namespace InternetOfVoice\VSMS\Core\Helper;

use \DateInterval;
use \DateTime;
use \Exception;

/**
 * Class SkillHelper
 *
 * @author  Alexander Schmidt <a.schmidt@internet-of-voice.de>
 * @license http://opensource.org/licenses/MIT
 */
class SkillHelper {
	/**
	 * getDateFromSeason
	 *
	 * Helper function to get a (start) date for a given year and season.
	 *
	 * Dates are not always accurate for northern and southern hemispheres, as
	 * solstices/equinoxes may vary by -1/+1 day.
	 *
	 * Seasons spanning two years (winter in northern and summer in southern
	 * hemisphere) might point to either current or last year. Current year
	 * is enforced by parameter $force_year, else the seasons point to last year.
	 *
	 * Based on code by http://biostall.com/get-the-current-season-using-php/
	 *
	 * @param   int     $year       Year
	 * @param   string  $season     Season {spring, summer, fall, winter}
	 * @param   string  $hemisphere Hemisphere {northern, southern, australia}
	 * @param   bool    $force_year Enforce given year?
	 *
	 * @return  false|DateTime
	 * @access  public
	 * @author  a.schmidt@internet-of-voice.de
	 */
    public function getDateFromSeason($year, $season, $hemisphere, $force_year = true) {
        $date   = false;
        $offset = $force_year ? 0 : 1;

        try {
	        switch($hemisphere) {
	            case 'northern':
	                switch($season) {
	                    case 'spring':
	                        $date = new DateTime($year . '-03-21');
	                    break;

	                    case 'summer':
	                        $date = new DateTime($year . '-06-21');
	                    break;

	                    case 'fall':
	                        $date = new DateTime($year . '-09-23');
	                    break;

	                    case 'winter':
	                        $date = new DateTime(($year - $offset) . '-12-21');
	                    break;
	                }
	            break;

	            case 'southern':
	                switch($season) {
	                    case 'fall':
	                        $date = new DateTime($year . '-03-21');
	                    break;

	                    case 'winter':
	                        $date = new DateTime($year . '-06-21');
	                    break;

	                    case 'spring':
	                        $date = new DateTime($year . '-09-23');
	                    break;

	                    case 'summer':
	                        $date = new DateTime(($year - $offset) . '-12-21');
	                    break;
	                }
	            break;

	            case 'australia':
	                switch($season) {
	                    case 'fall':
	                        $date = new DateTime($year . '-03-01');
	                    break;

	                    case 'winter':
	                        $date = new DateTime($year . '-06-01');
	                    break;

	                    case 'spring':
	                        $date = new DateTime($year . '-09-01');
	                    break;

	                    case 'summer':
	                        $date = new DateTime(($year - $offset) . '-12-01');
	                    break;
	                }
	            break;
	        }
        } catch(Exception $e) {
            // still false
        }

        if($date instanceof DateTime) {
            $date->setTime(0, 0, 0);
        }

        return $date;
    }

    /**
     * extractAmazonDate
     *
     * Extracts an AMAZON.DATE slot value to two DateTime objects 'start' and 'end', returned as [start, end]
     * If given date matches no date, both will be false. If it refers to "now", both will be the same DateTime.
     * In all other cases, start and end describe a time span corresponding to given date, e.g.
     * 2017-01-01 -> [2017-01-01 00:00:00, 2017-01-01 23:59:59]
     * 2017-05    -> [2017-05-01 00:00:00, 2017-05-31 23:59:59]
     *
     * Unspecific dates like "May" are always sent as future dates (next May) by Amazon.
     * You may use the optional $date_back parameter to enforce past dates (last May).
     * If we are right now in May, no backdating is performed.
     * Caution: please consider that a user might have actually referred to a future date.
     *
     * @param   string      $amazon_date    AMAZON.DATE slot value
     * @param   bool        $date_back      Enforce a date in the past?
     * @param   string      $hemisphere     'northern' (default), 'southern' or 'australia' - used for extracting a season date (ex: 2019-WI)
     * @return  DateTime[]
     * @access  public
     * @author  a.schmidt@internet-of-voice.de
     * @see     https://developer.amazon.com/public/solutions/alexa/alexa-skills-kit/docs/built-in-intent-ref/slot-type-reference#date
     */
    public function extractAmazonDate($amazon_date, $date_back = false, $hemisphere = 'northern') {
		$start = false;
		$end   = false;

		try {
	        switch (true) {
	            case ($amazon_date == 'PRESENT_REF'):
	                $origin = 'now';
	                $start  = new DateTime();
	                $end    = clone $start;
	            break;

	            // Date
	            case preg_match('~^[\d]{4}-[\d]{2}-[\d]{2}$~', $amazon_date):
	                $origin = 'date';
	                $start  = DateTime::createFromFormat('Y-m-d', $amazon_date);
	                $start->setTime(0, 0, 0);
	                $end = clone $start;
	                $end->setTime(23, 59, 59);
	            break;

		        // Day
		        case preg_match('~^XXXX-XX-([\d]{2})$~', $amazon_date, $matches):
			        $origin = 'day';
			        $start  = DateTime::createFromFormat('Y-m-d', date('Y') . '-' . date('m') . '-' . $matches[1]);
			        $start->setTime(0, 0, 0);
			        $end = clone $start;
			        $end->setTime(23, 59, 59);
		        break;

		        // Week
	            case preg_match('~^([\d]{4})-W([\d]{1,2})$~', $amazon_date, $matches):
	                $origin = 'week';
	                $start  = new DateTime();
	                $start->setTime(0, 0, 0);
	                $start->setISODate(intval($matches[1]), intval($matches[2]));
	                $end = clone $start;
	                $end->add(new DateInterval('P1W'))->sub(new DateInterval('PT1S'));
	            break;

	            // Weekend
	            case preg_match('~^([\d]{4})-W([\d]{1,2})-WE$~', $amazon_date, $matches):
	                $origin = 'weekend';
	                $start  = new DateTime();
	                $start->setTime(0, 0, 0);
	                $start->setISODate(intval($matches[1]), intval($matches[2]));
	                $start->add(new DateInterval('P5D'));
	                $end = clone $start;
	                $end->add(new DateInterval('P2D'))->sub(new DateInterval('PT1S'));
	            break;

	            // Month
	            case preg_match('~^([\d]{4}-[\d]{2})$~', $amazon_date, $matches):
	            case preg_match('~^([\d]{4}-[\d]{2})-XX$~', $amazon_date, $matches):
	                $origin = 'month';
	                $start  = DateTime::createFromFormat('Y-m-d', $matches[1] . '-01');
	                $start->setTime(0, 0, 0);
	                $end = clone $start;
	                $end->add(new DateInterval('P1M'))->sub(new DateInterval('PT1S'));
	            break;

	            // Year
	            case preg_match('~^([\d]{4})$~', $amazon_date, $matches):
	            case preg_match('~^([\d]{4})-XX-XX$~', $amazon_date, $matches):
	                $origin = 'year';
	                $start  = DateTime::createFromFormat('Y-m-d', $matches[1] . '-01-01');
	                $start->setTime(0, 0, 0);
	                $end = clone $start;
	                $end->add(new DateInterval('P1Y'))->sub(new DateInterval('PT1S'));
	            break;

	            // Decade
	            case preg_match('~^([\d]{3})X$~', $amazon_date, $matches):
	                $origin = 'decade';
	                $start  = DateTime::createFromFormat('Y-m-d', $matches[1] . '0-01-01');
	                $start->setTime(0, 0, 0);
	                $end = clone $start;
	                $end->add(new DateInterval('P10Y'))->sub(new DateInterval('PT1S'));
	            break;

	            // Season
	            case preg_match('~^([\d]{4})-(SP|SU|FA|WI)$~', $amazon_date, $matches):
	                $seasons = ['SP' => 'spring', 'SU' => 'summer', 'FA' => 'fall', 'WI' => 'winter'];
	                $origin  = 'season';
	                $start   = $this->getDateFromSeason(intval($matches[1]), $seasons[$matches[2]], $hemisphere, false);
	                $end     = clone $start;
	                $end->add(new DateInterval('P3M'))->sub(new DateInterval('PT1S'));
	            break;

	            default:
	                // Unknown format
	                return [$start, $end];
	            break;
	        }

	        if($start !== false && $date_back) {
		        $now  = new DateTime();
		        $diff = $now->getTimestamp() - $start->getTimestamp();
		        if($diff < 0) {
			        if(in_array($origin, ['date', 'month', 'season'])) {
				        // Need to date back as requested with $date_back for some origins.
				        if ($diff > - 604800) {
					        // If difference is less than seven days and origin is date, most likely a day name was given.
					        $start->sub(new DateInterval('P7D'));
					        $end->sub(new DateInterval('P7D'));
				        } else {
					        // In all other cases, subtract a year (for dates like "3rd of November", months like "August"
					        // and seasons like "Winter")
					        $start->sub(new DateInterval('P1Y'));
					        $end->sub(new DateInterval('P1Y'));
				        }
			        }

			        if(in_array($origin, ['day'])) {
			        	// As a day number was given, subtract one month
				        $start->sub(new DateInterval('P1M'));
				        $end->sub(new DateInterval('P1M'));
			        }
		        }
	        }
		} catch(Exception $e) {
			// still false, false
		}

        return [$start, $end];
    }

    /**
     * extractAmazonTime
     *
     * Extracts an AMAZON.TIME slot value and sets time portion of a given date (or today, if omitted)
     *
     * For time period indicators, the following times are returned:
     * night: 00:00, morning: 06:00, afternoon: 12:00, evening: 18:00
     *
     * @param   string      $amazon_time        AMAZON.TIME slot value
     * @param   DateTime    $date               DateTime object or null
     *
     * @return  false|DateTime
     * @access  public
     * @author  a.schmidt@internet-of-voice.de
     * @see     https://developer.amazon.com/public/solutions/alexa/alexa-skills-kit/docs/built-in-intent-ref/slot-type-reference#time
     */
    public function extractAmazonTime($amazon_time, $date = null) {
		try {
	        if(!$date instanceof DateTime) {
	            $date = new DateTime;
	        }

	        switch(true) {
	            case preg_match('~^(\d{2}:\d{2})~', $amazon_time, $matches):
	                list($hours, $minutes) = explode(':', $matches[1]);
	                $date->setTime($hours, $minutes, 0);
	            break;

	            case ($amazon_time == 'NI'):
	                $date->setTime(0, 0, 0);
	            break;

	            case ($amazon_time == 'MO'):
	                $date->setTime(6, 0, 0);
	            break;

	            case ($amazon_time == 'AF'):
	                $date->setTime(12, 0, 0);
	            break;

	            case ($amazon_time == 'EV'):
	                $date->setTime(18, 0, 0);
	            break;

		        default:
		            // do nothing, return defaults set above
	            break;
	        }
		} catch(Exception $e) {
			$date = false;
		}

	    return $date;
    }


    /**
     * convertDateTimeToHuman
     *
     * Converts a DateTime object to a speakable text.
     *
     * Omits year, if date is current year and omits minutes, if zero (if option 'with_time' is set).
     * Can be configured to reply with relative day names (today, yesterday, friday, ...) instead of day number
     *
     * Options array values:
     * relative:     return "today", "yesterday", "tomorrow" or "friday" if applicable; day number otherwise
     * with_time:    including time portion in response (seconds are omitted)
     * with_year:    enforce inclusion of year (current year is omitted by default)
     * with_minutes: enforce inclusion of minutes (zero minute is omitted by default)
     *
     * Translations array keys:
     * days:        ['Monday', ..., 'Sunday']
     * months:      ['January', ..., 'December']
     * today:       'today'
     * yesterday:   'yesterday'
     * tomorrow:    'tomorrow'
     * prefix_48h:  prefix for today / yesterday / tomorrow (like 'on' or 'at' or empty string)
     * prefix_rel:  prefix for day names (like 'on' or 'at' or empty string) -- deprecated
     * prefix_last: prefix for day names in history (like 'last' or empty string)
     * prefix_next: prefix for day names in future (like 'next' or empty string)
     * prefix_abs:  prefix for absolute dates (like 'on' or 'at' or empty string)
     * at:          'at'
     * o'clock:     'o\'clock' (mind proper apostrophe escaping)
     *
     * @param   DateTime        $date           DateTime object
     * @param   array           $options        Options
     * @param   array           $translations   Translations
     * @return  string
     * @access  public
     * @author  a.schmidt@internet-of-voice.de
     */
    public function convertDateTimeToHuman(DateTime $date, $options = [], $translations = []) {
	    $result = array();

		try {
		    $now  = new DateTime();
		    $diff = $now->getTimestamp() - $date->getTimestamp();

	        // backwards compatibility
	        if(isset($translations['prefix_rel'])) {
	            $translations['prefix_last'] = isset($translations['prefix_last']) ? $translations['prefix_last'] : $translations['prefix_rel'];
	            $translations['prefix_next'] = isset($translations['prefix_next']) ? $translations['prefix_next'] : $translations['prefix_rel'];
	        }

	        if(in_array('relative', $options) && ($diff > 0) && ($diff < 604800)) {
	            // Relative day, if not more than 7 days ago
	            // Today
	            if($date->format('d') == date('d')) {
		            array_push($result, $translations['prefix_48h']);
	                array_push($result, $translations['today']);

	            // Yesterday
	            } elseif($date->format('d') == date('d', time() - 86400)) {
		            array_push($result, $translations['prefix_48h']);
	                array_push($result, $translations['yesterday']);

	            // Day name
	            } else {
		            array_push($result, $translations['prefix_last']);
	                array_push($result, $translations['days'][$date->format('N') - 1]);
	            }
	        } elseif(in_array('relative', $options) && ($diff <= 0) && ($diff > -604800)) {
	            // Relative day, if not more than 7 days in future
	            // Today
	            if ($date->format('d') == date('d')) {
	                array_push($result, $translations['prefix_48h']);
	                array_push($result, $translations['today']);

	            // Yesterday
	            } elseif ($date->format('d') == date('d', time() + 86400)) {
	                array_push($result, $translations['prefix_48h']);
	                array_push($result, $translations['tomorrow']);

	            // Day name
	            } else {
	                array_push($result, $translations['prefix_next']);
	                array_push($result, $translations['days'][$date->format('N') - 1]);
	            }
	        } else {
	            array_push($result, $translations['prefix_abs']);

	            // Day without leading zero, including enumeration point
	            array_push($result, $date->format('j.'));

	            // Month name
	            array_push($result, $translations['months'][$date->format('n') - 1]);

	            // Year
	            $year = $date->format('Y');
	            if($year != date('Y') || in_array('with_year', $options)) {
	                array_push($result, $year);
	            }
	        }

	        if(in_array('with_time', $options)) {
	            array_push($result, $translations['at']);

	            // Hour without leading zero
	            array_push($result, $date->format('G'));
	            array_push($result, $translations['o\'clock']);

	            // Minutes without leading zero
	            $minutes = intval($date->format('i'));
	            if($minutes > 0  || in_array('with_minutes', $options)) {
	                array_push($result, $minutes);
	            }
	        }
		} catch(Exception $e) {
			// still empty array
		}

        $result = trim(implode(' ', $result));
		if(empty($result)) {
			return false;
		} else {
			return preg_replace('/\s+/', ' ', $result);

		}
    }

    /**
     * Converts a list to a comma separated string and using a conjunction word before the last element
     * Example: [one, two, three] -> one, two and three
     *
     * @param   array       $list
     * @param   string      $conjunction
     * @return  string
     * @access  public
     * @author  a.schmidt@internet-of-voice.de
     */
    public function convertListToHuman($list, $conjunction = 'and') {
        if(count($list) > 1) {
            $last1 = array_pop($list);
            $last2 = count($list) - 1;
            $list[$last2] = $list[$last2] . ' ' . $conjunction . ' ' . $last1;
        }

        return implode(', ', $list);
    }
}
